src/libraries/System.Threading.Channels/src/System/Threading/Channels/ChannelUtilities.cs

```diff
@@ -66,60 +66,262 @@ internal static ValueTask<T> GetInvalidCompletionValueTask<T>(Exception error)
+                    Debug.Assert(current is not null);
+
+                    TAsyncOp? next = current.Next;
+                    Debug.Assert(next is not null);
 
-                    bool completed = error != null ? c.TrySetException(error) : c.TrySetResult(result);
-                    Debug.Assert(completed || c.CancellationToken.CanBeCanceled);
+                    current.Next = current.Previous = null;
 
-                    c = next;
+                    current.TrySetResult(result);
+
+                    current = next;
                 }
-                while (c != head);
+                while (current != head);
+
+                head = null;
             }
         }
 
-        /// <summary>Removes all operations from the queue, failing each.</summary>
-        /// <param name="operations">The queue of operations to complete.</param>
+        /// <summary>Iterates through the linked list, successfully completing each operation that should have already had completion reserved.</summary>
+        /// <param name="head">The head of the queue of operations to complete.</param>
+        /// <param name="result">The result with which to complete each operations.</param>
+        internal static void DangerousSetOperations<TAsyncOp, TResult>(TAsyncOp? head, TResult result)
+            where TAsyncOp : AsyncOperation<TAsyncOp, TResult>
+        {
+            TAsyncOp? current = head;
+            if (current is not null)
+            {
+                do
+                {
+                    Debug.Assert(current is not null);
+
+                    TAsyncOp? next = current.Next;
+                    Debug.Assert(next is not null);
+
+                    current.Next = current.Previous = null;
+
+                    current.DangerousSetResult(result);
+
+                    current = next;
+                }
+                while (current != head);
+            }
+        }
+
+        /// <summary>Iterates through the linked list, reserving completion for every element.</summary>
+        /// <param name="head">The head of the queue of operations to complete.</param>
+        /// <returns>A linked list of all successfully reserved operations. All other operations are ignored and dropped.</returns>
+        internal static TAsyncOp? TryReserveCompletionIfCancelable<TAsyncOp>(ref TAsyncOp? head)
+            where TAsyncOp : AsyncOperation<TAsyncOp>
+        {
+            TAsyncOp? reserved = null;
+
+            TAsyncOp? current = head;
+            if (current is not null)
+            {
+                do
+                {
+                    Debug.Assert(current is not null);
+
+                    TAsyncOp? next = current.Next;
+                    Debug.Assert(next is not null);
+
+                    current.Next = current.Previous = null;
+
+                    if (current.TryReserveCompletionIfCancelable())
+                    {
+                        Enqueue(ref reserved, current);
+                    }
```

Comment:
The common case can be optimized to be a nop on the list.

```cs
if (!current.TryReserveCompletionIfCancelable())
{
    Remove(ref head, current);
}
```

Reply:
I'd done it the way I did to avoid needing to keep track of the list invariants within the loop. It's also not much of an optimization. But sure.

Reply:
Just took another look at it, and calling Remove in the loop makes it really hard to reason about, e.g. Remove might set head to null, which then causes the outer loop to be infinite unless special additional care is taken. I'm going to leave it as it is.